src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -144,415 +143,344 @@ void TemporalMemory::initialize(
+             vector<SegmentOverlap>::const_iterator activeSegmentsEnd,
+             vector<SegmentOverlap>::const_iterator matchingSegment,
+             vector<SegmentOverlap>::const_iterator matchingSegmentsEnd,
+             UInt cellsPerColumn)
+      :activeColumn_(activeColumn),
+       activeColumnsEnd_(activeColumnsEnd),
+       activeSegment_(activeSegment),
+       activeSegmentsEnd_(activeSegmentsEnd),
+       matchingSegment_(matchingSegment),
+       matchingSegmentsEnd_(matchingSegmentsEnd),
+       cellsPerColumn_(cellsPerColumn),
+       finished_(false)
     {
-      UInt column = columnForCell(cell);
-
-      if (activeColumns.find(column) == activeColumns.end())
-      {
-        _predictedInactiveCells.insert(cell);
-      }
+      calculateNext_();
     }
-  }
-
-  return make_tuple(_activeCells, _winnerCells,
-    _predictedColumns, _predictedInactiveCells);
-}
-
-tuple<set<Cell>, set<Cell>, vector<Segment>> TemporalMemory::burstColumns(
-  set<UInt>& activeColumns,
-  set<UInt>& predictedColumns,
-  set<Cell>& prevActiveCells,
-  set<Cell>& prevWinnerCells,
-  Connections& _connections)
-{
-  set<Cell> _activeCells;
-  set<Cell> _winnerCells;
-  vector<Segment> _learningSegments;
-
-  vector<UInt> _unpredictedColumns(activeColumns.begin(), activeColumns.end());
-
-  if (predictedColumns.size() > 0)
-  {
-    // Resize to the worst case usage
-    _unpredictedColumns.resize(activeColumns.size() + predictedColumns.size());
-
-    // Remove the predicted columns from the 
-    // currently active columns
-    vector<UInt>::iterator it = set_difference(
-      activeColumns.begin(), activeColumns.end(),
-      predictedColumns.begin(), predictedColumns.end(),
-      _unpredictedColumns.begin());
-
-    // Trim remainer of set
-    _unpredictedColumns.resize(it - _unpredictedColumns.begin());
-  }
-
-  // Sort unpredictedActiveColumns before iterating for python compatibility
-  sort(_unpredictedColumns.begin(), _unpredictedColumns.end());
-
-  for (Int column : _unpredictedColumns)
-  {
-    Segment bestSegment;
-    Cell bestCell;
-    bool foundCell = false;
-    bool foundSegment = false;
-
-    vector<Cell> cells = cellsForColumnCell(column);
 
-    for (auto cell : cells)
-      _activeCells.insert(cell);
-
-    tie(foundCell, bestCell, foundSegment, bestSegment) =
-      bestMatchingCell(cells, prevActiveCells, _connections);
-
-    if (foundCell)
+    bool operator !=(const Iterator& other)
     {
-      _winnerCells.insert(bestCell);
+      return finished_ != other.finished_ ||
+        activeColumn_ != other.activeColumn_ ||
+        activeSegment_ != other.activeSegment_ ||
+        matchingSegment_ != other.matchingSegment_;
     }
 
-    if (!foundSegment && prevWinnerCells.size())
+    const ExcitedColumnData& operator*() const
     {
-      bestSegment = _connections.createSegment(bestCell);
-      foundSegment = true;
+      NTA_ASSERT(!finished_);
```

Comment:
Correct. The only way that could happen is if someone in this file screws up, so I think an assert makes sense. I think of NTA_CHECK as being more for checking parameters of the public interface.